pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java

```diff
@@ -234,18 +298,34 @@ private FunctionContext convertRexExpressionsToFunctionContext(RexExpression.Fun
     return functionContext;
   }
 
+  private void rewriteAggArgumentWithLiterals(List<ExpressionContext> aggArguments, int aggIdx,
+      RexExpression.FunctionCall aggFunctionCall) {
+    String functionName = aggFunctionCall.getFunctionName();
+    // This should be provided by AggregationFunctionType indicating which argIdx requires literal.
+    if (functionName.equals("FIRSTWITHTIME") || functionName.equals("LASTWITHTIME")) {
+      aggArguments.add(ExpressionContext.forIdentifier("__PLACEHOLDER__"));
+      aggArguments.add(ExpressionContext.forLiteralContext(_aggCallLiteralArgsMap.get(aggIdx).get(2)));
+    }
+    if (functionName.equals("PERCENTILE")) {
+      aggArguments.add(ExpressionContext.forLiteralContext(_aggCallLiteralArgsMap.get(aggIdx).get(1)));
+    }
+  }
```

Comment:
this function needs to be replaced by a util in `AggregateFunctionType` that handles each registered function separately.